mesonpy/_editable.py

```diff
@@ -0,0 +1,153 @@
+    trying to load our modules, and trigger a rebuild. After triggering a
+    rebuild, we return None in find_spec, letting the normal finders pick up the
+    modules.
+    """
+
+    def __init__(
+        self,
+        project_path: str,
+        build_path: str,
+        import_paths: List[str],
+        top_level_modules: List[str],
+        rebuild_commands: List[List[str]],
+        verbose: bool = False,
+    ) -> None:
+        self._project_path = project_path
+        self._build_path = build_path
+        self._import_paths = import_paths
+        self._top_level_modules = top_level_modules
+        self._rebuild_commands = rebuild_commands
+        self._verbose = verbose
+
+    def __repr__(self) -> str:
+        return f'{self.__class__.__name__}({self._project_path})'
+
+    def _debug(self, msg: str) -> None:
+        if self._verbose:
+            print(msg.format(**_STYLES))
+
+    def _proc(self, command: List[str]) -> None:
+        # skip editable hook installation in subprocesses, as during the build
+        # commands the module we are rebuilding might be imported, causing a
+        # rebuild loop
+        # see https://github.com/mesonbuild/meson-python/pull/87#issuecomment-1342548894
+        env = os.environ.copy()
+        env['_MESONPY_EDITABLE_SKIP'] = os.pathsep.join((
+            env.get('_MESONPY_EDITABLE_SKIP', ''),
+            self._project_path,
+        ))
+
+        if self._verbose:
+            subprocess.check_call(command, cwd=self._build_path, env=env)
+        else:
+            subprocess.check_output(command, cwd=self._build_path, env=env)
+
+    @functools.lru_cache(maxsize=1)
+    def rebuild(self) -> None:
+        self._debug(f'{{cyan}}{{bold}}+ rebuilding {self._project_path}{{reset}}')
+        for command in self._rebuild_commands:
+            self._proc(command)
+        self._debug('{cyan}{bold}+ successfully rebuilt{reset}')
+
+    def find_spec(
+        self,
+        fullname: str,
+        path: Optional[Sequence[Union[str, bytes]]],
+        target: Optional[ModuleType] = None,
+    ) -> None:
+        # if it's one of our modules, trigger a rebuild
+        if fullname.split('.', maxsplit=1)[0] in self._top_level_modules:
+            self.rebuild()
+            # prepend the project path to sys.path, so that the normal finder
+            # can find our modules
+            # we prepend so that our path comes before the current path (if
+            # the interpreter is run with -m), see gh-239
+            if sys.path[:len(self._import_paths)] != self._import_paths:
+                for path in self._import_paths:
+                    if path in sys.path:
+                        sys.path.remove(path)
+                sys.path = self._import_paths + sys.path
+        # return none (meaning we "didn't find" the module) and let the normal
+        # finders find/import it
+        return None
+
+    @classmethod
+    def install(
```

Comment:
This function needs two changes:
- add a `self._debug(f'{{cyan}}{{bold}}+` print command like for `rebuild` above
- call `meson install --only-changed` rather than reinstalling all files

Reply:
actually, inside the `.rebuild` method I think. It happens before the `+ successfully rebuilt` message.

Reply:
This method only installs the hook, the meson install is part of the rebuild process. It seems we lost `--only-changed` in the refactoring, will correct.

Reply:
Are you sure ``--only-changed`` is working?

What if I delete a file from the source directory(using ``install_subdir`` to install files)? This doesn't seem to be reflected(the file is not deleted from the install directory).

Reply:
`meson install` doesn't delete files that were installed by older versions of the project but then deleted from source.

`--only-changed` has no effect on that -- it just ***avoids doing I/O on files that would otherwise be installed***, if the destination already exists and the heuristic of "file timestamps" indicates that the source isn't newer than the destination.

Reply:
I don't think this is a huge issue, but definitely something we'll want to fix.

@eli-schwartz would it be reasonable to request this as a feature from Meson? Or should we just implement it ourselves.
`--only-changed` not removing deleted files seems like a footgun to me.

Reply:
If the file was deleted, then "probably" any actual references to it inside another file were deleted as well, so having an extra pointless file may not matter.

Although sometimes it does matter -- for example, if you rename `mesonbuild/mesonlib/*.py` to `mesonbuild/utils/*.py` and then add `mesonbuild/mesonlib.py` with the old contents of `mesonbuild/mesonlib/__init__.py` (`from .foo import *` remapped to `.utils.foo`). And then you re-run `python -m build` to produce wheels, and upload them to PyPI, but setuptools preserves the old code in `build/`, does not clean it up, and the wheel is broken because `mesonbuild.mesonlib` is still used, it just changed from a package to a module.

(Guess how I know that one.)

Meson doesn't have that problem if you install fresh as part of making a brand new wheel. Or if you install fresh to a DESTDIR as part of a unix packaging workflow. Or if you run `ninja uninstall && ninja install` to delete the old version and install the new one.

But it does have that problem -- it *has* to have that problem -- if you simply install a newer version in-place on top of an older version.

Reply:
@FFY00,

I think having `meson install --only-changed` produce results that are intentionally different from `meson install` is probably wrong. The point of this option is to say "I'm pretty sure I don't have timestamp bugs such as changed files that don't have updated modification times, so go ahead and optimize the work when you're positive that it doesn't make a difference".

It sounds like what you want is better named `--uninstall-old`.

Reply:
Yes, that is why it matters, and also why I don't think it's a *huge* issue, but definitely something we want to fix :blush: 

> But it does have that problem -- it _has_ to have that problem -- if you simply install a newer version in-place on top of an older version.

It could keep track of the installed files in a metadata file, and when installing with `--only-changed`, it would look at it, delete any files not present in the new installation, and update the metadata file.

Does this sound reasonable? I can open a Meson issue if you think it's worth considering. Otherwise, we can implement that here ourselves, but it IMO being a footgun for `--only-changed` in Meson, I'd rather see it fixed in the source directly.

Reply:
(sorry, when I was writing the message above the UI hadn't updated with your last reply)

> I think having `meson install --only-changed` produce results that are intentionally different from `meson install` is probably wrong. The point of this option is to say "I'm pretty sure I don't have timestamp bugs such as changed files that don't have updated modification times, so go ahead and optimize the work when you're positive that it doesn't make a difference".

I personally disagree, but understand your position.

> It sounds like what you want is better named `--uninstall-old`.

That seems reasonable.

Reply:
I guess it does seem reasonable for meson not to touch other files.

We also probably can figure out which files are changed through introspection, as long as intro-install-plan and intro-installed match what happens without ``--only-changed``. (Then, we find a list of the installed files, and delete everything else in the folder not in installed files).